code/game/gamemodes/umbrage/umbrage_datum.dm

```diff
@@ -0,0 +1,136 @@
+//Handles everything related to psi and lucidity.
+/datum/umbrage
+	var/datum/mind/linked_mind
+	var/mob/living/linked_body
+	var/psi = 100 //Our psi, used for abilities.
+	var/max_psi = 100
+	var/psi_regeneration = 20 //The limit for psi regenerated during the cycle.
+	var/psi_used_since_last_cycle //How much psi we've used in the last five seconds.
+	var/cycle_progress = 0 //When this reaches 5, it will reset to 0 and regenerate up to 20 spent psi.
+	var/lucidity = 3 //Lucidiy is used to buy abilities. We gain one each time we drain someone.
+	var/lucidity_drained = 3 //How much lucidity we've taken overall.
+	var/list/linked_abilities = list() //Abilities that we have
+	var/list/linked_ability_types = list() //Used for cloning and gaining new bodies
+	var/list/drained_minds = list() //Minds eaten with Devour Will.
+
+/datum/umbrage/New()
+	..()
+	START_PROCESSING(SSprocessing, src)
+
+/datum/umbrage/process()
+	cycle_progress++
+	if(cycle_progress >= 5)
+		regenerate_psi()
+		cycle_progress = 0
+	if(linked_mind && linked_body != linked_body && linked_body.stat != DEAD)
+		link_to_new_body(linked_body)
+
+/datum/umbrage/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	return ..()
+
+/datum/umbrage/proc/link_to_new_body(mob/living/new_host) //Give all our old abilities to a new body, in case we're cloned or such
+	linked_body = new_host
+	for(var/taip in linked_ability_types)
+		var/datum/action/innate/umbrage/U = taip
+		give_ability(initial(U.id), 1)
+	return 1
+
+/datum/umbrage/proc/use_psi(used_psi)
+	cycle_progress = 0 //Reset regenerating psi when we use more
+	psi = max(0, min(psi - used_psi, max_psi))
+	psi_used_since_last_cycle += used_psi
+	return 1
+
+/datum/umbrage/proc/regenerate_psi()
+	var/psi_to_regen
+	if(psi >= max_psi) //No need to regenerate anything, so let's get out
+		return 1
+	psi_to_regen = min(psi_used_since_last_cycle, psi_regeneration) //Never go above our regen rate
+	psi = min(psi + psi_to_regen, max_psi)
+	psi_used_since_last_cycle = 0
+	return 1
+
+/datum/umbrage/proc/has_ability(id)
+	for(var/datum/action/innate/umbrage/U in linked_abilities)
+		if(U.id == id)
+			return 1
+	return
+
+/datum/umbrage/proc/give_ability(id, silent, consume_lucidity) //Gives an ability of a certain name to the umbrage and consumes lucidity if applicable.
+	var/datum/action/innate/umbrage/ability
+	for(var/V in subtypesof(/datum/action/innate/umbrage))
+		var/datum/action/innate/umbrage/U = V
+		if(initial(U.id) == id)
```

Comment:
Okay, I've been having some trouble with this. Here's what I have so far:
```
for(var/V in subtypesof(/datum/action/innate/umbrage))
		var/datum/action/innate/umbrage/U = V
		ability_types[initial(U.id)] = U
```
This doesn't work. The problem is that I've thus far been unable to use the type variable. I can't have initial(U.type) because it never changes and so it can't use initial. I can't use U.type normally because it's not a defined variable, and I can't use U. Any ideas?